chain/chain/src/missing_chunks.rs

```diff
@@ -0,0 +1,292 @@
+use log::warn;
+use near_primitives::hash::CryptoHash;
+use near_primitives::sharding::ChunkHash;
+use near_primitives::types::BlockHeight;
+use std::cmp::Ordering;
+use std::collections::{
+    btree_map::{self, BTreeMap},
+    hash_map::{self, HashMap},
+    BinaryHeap, HashSet,
+};
+
+type BlockHash = CryptoHash;
+
+const MAX_BLOCKS_MISSING_CHUNKS: usize = 1024;
+
+pub trait BlockLike {
+    fn hash(&self) -> BlockHash;
+    fn height(&self) -> BlockHeight;
+}
+
+#[derive(Debug)]
+struct HeightOrdered<T>(T);
+
+impl<T: BlockLike> PartialEq for HeightOrdered<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.height() == other.0.height()
+    }
+}
+impl<T: BlockLike> Eq for HeightOrdered<T> {}
+impl<T: BlockLike> PartialOrd for HeightOrdered<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.0.height().cmp(&other.0.height()))
+    }
+}
+impl<T: BlockLike> Ord for HeightOrdered<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.height().cmp(&other.0.height())
+    }
+}
+
+/// Structure for keeping track of missing chunks.
+/// The reason to have a Block type parameter instead of using the
+/// `block::Block` type is to make testing easier (`block::Block` is a complex structure and I
+/// don't care about most of it).
+#[derive(Debug, Default)]
+pub struct MissingChunksPool<Block: BlockLike> {
+    missing_chunks: HashMap<ChunkHash, HashSet<BlockHash>>,
+    blocks_missing_chunks: HashMap<BlockHash, HashSet<ChunkHash>>,
+    blocks_waiting_for_chunks: HashMap<BlockHash, Block>,
+    blocks_ready_to_process: BinaryHeap<HeightOrdered<Block>>,
+    height_idx: BTreeMap<BlockHeight, HashSet<BlockHash>>,
+}
+
+impl<Block: BlockLike> MissingChunksPool<Block> {
+    pub fn new() -> Self {
```

Comment:
Yes, I originally was deriving `Default`, but when I switched to using a type parameter it meant that the type you use also needs to have a `Default` implementation (which is dumb because I don't need any instances of `Block` to initialize the data structure). Since there is no `Default` implementation for `block::Block` I needed to put in the `new` boilerplate.